docs: migrate from 0.9 guide (#1873)
firebase/genkit: docs/migrating-from-0.9.md

> # Migrating from 0.9

Reviewer: I think we need to mention the authPolicy change, and ToolDefinition being removed, the dotprompt  package change

Reviewer 2: I think it would be helpful to discuss the new context concept in a bit more detail too. It seems like auth policy is a part of that, but it feels like an overall new concept that warrants some attention.

Author: Done.

> Genkit 1.0 introduces some of breaking changes alongside feature enhancements that improve overall functionality. If you have been developing applications with Genkit 0.9, you will need to update your application code when you upgrade to the latest version. This guide outlines the most significant changes and offers steps to migrate your existing applications smoothly.

Reviewer: ```suggestion
Genkit 1.0 introduces many feature enhancements that improve overall functionality as well as some some breaking changes. If you have been developing applications with Genkit 0.9, you will need to update your application code when you upgrade to the latest version. This guide outlines the most significant changes and offers steps to migrate your existing applications smoothly.
```

> Genkit 1.0 introduces many feature enhancements that improve overall functionality as well as some some breaking changes. If you have been developing applications with Genkit 0.9, you will need to update your application code when you upgrade to the latest version. This guide outlines the most significant changes and offers steps to migrate your existing applications smoothly.

Author: ```suggestion
Genkit 1.0 introduces many feature enhancements that improve overall functionality as well as some breaking changes. If you have been developing applications with Genkit 0.9, you will need to update your application code when you upgrade to the latest version. This guide outlines the most significant changes and offers steps to migrate your existing applications smoothly.
```

> If you're using 0.5, use <> to upgrade to 0.9 first before upgrading to 1.0

Reviewer: ```suggestion
If you're using 0.5, use the [0.5 to 0.9 migration guide](https://firebase.google.com/docs/genkit/migrating-from-0.5) to upgrade to 0.9 first before upgrading to 1.0
```

> ## Streaming functions do not require an await

Reviewer: I suggest ordering these in most largest conceptual order first, so I'd start with the Beta API before the more point-wise changes.

**Old:**

```ts
const { stream, response } = await ai.generateStream(`hi`);
const { stream, output } = await myflow.stream(`hi`);
```

**New**

```ts
const { stream, response } = ai.generateStream(`hi`);
const { stream, output } = myflow.stream(`hi`);
```

## Beta APIs

> We're introducing an unstable, Beta API channel, and leaving session, chat and Genkit client APIs in beta as we continue to refine them. 

Reviewer: I think we need to be explicit here about `createSession` and `chat` as the nouns, so people can control-f in this doc and find the fix for any errors

Author: Done.

Note: When using the APIs as part of the Beta API, you may experience breaking changes outside of SemVer. Breaking chanhes may occur on minor release (we'll try to avoid making breaking beta API changes on patch releases).

**Old:**

```ts
import { genkit } from 'genkit';
const ai = genkit({...})
const session = ai.createSession({ ... })
```

**New**

```ts
import { genkit } from 'genkit/beta';
const ai = genkit({...})
const session = ai.createSession({ ... })
```

**Old:**

```ts
import { runFlow, streamFlow } from 'genkit/client';
```

**New**

```ts
import { runFlow, streamFlow } from 'genkit/beta/client';
```

## Introducing new `@genkit-ai/express` package

> The package contains utilities to make it easier to build an express app using Express.js.

Reviewer: ```suggestion
This new package contains utilities to make it easier to build an express app using Express.js.
```

> Refer to https://js.api.genkit.dev/modules/_genkit-ai_express.html for more details.

Reviewer: I'd expect this to go to a github doc as the primary source

Author: we'll try to document more an more things in api reference...

Reviewer: it'll be a bit confusing to have 3 sources of truth, the API reference, the github repo and the Firebase dcos

Reviewer 3: By "the GitHub repo" do you mean the samples there or the docs in GitHub?

Reviewer: The docs in GitHub

Author: we want to prompte https://js.api.genkit.dev, I'm going to leave it for now.

> This new package contains utilities to make it easier to build an express app using Express.js.

Author: ```suggestion
This new package contains utilities to make it easier to build an Express.js server with Genkit.
```

> `startFlowServer` has moved from part of the genkit object to the express package, to use startFlowServer, update your imports. 

Reviewer: ```suggestion
`startFlowServer` has moved from part of the genkit object to this new `@genkit-ai/express` package, to use startFlowServer, update your imports. 
```

**Old:**

```ts
const ai = genkit({ ... });
ai.startFlowServer({
  flows: [myFlow1, myFlow2],
});
```

**New**

```ts
import { startFlowServer } from '@genkit-ai/express';
startFlowServer({
  flows: [myFlow1, myFlow2],
});
```

> ## Changes to Flows

Author: Good idea. Done.

> `run` function for custom trace blocks has moving to `ai.run`

Reviewer: ```suggestion
The `run` function for custom trace blocks has moved to part of the `genkit` object, invoke it with `ai.run` instead
```

Reviewer: This needs an old/new block

> **Old:**

Reviewer: Wed need to describe the change that matches `old` and `new` for the blocks below. Streaming callback becomes context and sendChunk?

Author: I talk about streaming callback just below.

```ts
ai.defineFlow({name: 'banana'}, async (input, streamingCallback) => {
  const step = await run('myCode', async () => {
    return 'something'
  });
})

> ```ts

Reviewer 2: remove the ts here, causing md compile issues

**New**

```ts
ai.defineFlow({name: 'banana'}, async (input, {context, sendChunk}) => {
  const step = await ai.run('myCode', async () => {
    return 'something'
  });
})
```

> `ai.defineStreamingFlow` has been removed. Use `ai.defineFlow` instead. Also, `streamingCallback` moved to a field inside the second argument of he flow function and is now called `sendChunk`.

Reviewer: The second section move above

Author: will keep it here... I removed streaming callback above

> `ai.defineStreamingFlow` has been removed. Use `ai.defineFlow` instead. Also, `streamingCallback` moved to a field inside the second argument of he flow function and is now called `sendChunk`.

Reviewer: Probably start the section with `ai.defineStreamingFlow` and show that it's a drop-in replacement to get `ai.defineFlow`

Author: it's technically not a drop-in replacement... I mean, it is for defineFlow, but not for the way to call the flow. That interface is different.

Reviewer: I guess, if someone was using `defineStreamingFlow` before, what steps do they need to take since it's going away, we should show that

**Old:**

```ts
const flow = ai.defineStreamingFlow({name: 'banana'}, async (input, streamingCallback) => {
  streamingCallback({chunk: 1});
})

const {stream} = await flow()
for await (const chunk of stream) {
  // ...
}
```

**New**

```ts
const flow = ai.defineFlow({name: 'banana'}, async (input, {context, sendChunk}) => {
  sendChunk({chunk: 1});
})

const {stream, output} = flow.stream(input);
for await (const chunk of stream) {
  // ...
}
```

> Flow auth is now called context. You can access auth as a field inside context:

Reviewer: This sentence doesn't match the example.  Maybe `flowAuth` is has been moved to be a field on a broader `context` object.   Instead of `getFlowAuth()` access the object as `context.auth`

**Old:**

```ts
ai.defineFlow({name: 'banana'}, async (input) => {
  const auth = getFlowAuth();
  // ...
})
```

**New**

```ts
ai.defineFlow({name: 'banana'}, async (input, { context }) => {
  const auth = context.auth;
})
```

> `onFlow` moved to `firebase-functions/https` package and got renamed to `onCallGenkit`. Here's how to use it:

Reviewer: This would benefit from an `old` version to show how this is different than the code would have been before

Author: done.

```ts
import { onCallGenkit } from "firebase-functions/https";
import { defineSecret } from "firebase-functions/params";
import { genkit, z } from "genkit";

const apiKey = defineSecret("GOOGLE_GENAI_API_KEY");

const ai = genkit({
  plugins: [googleAI()],
  model: gemini15Flash,
});

export const jokeTeller = ai.defineFlow(
  {
    name: "jokeTeller",
    inputSchema: z.string().nullable(),
    outputSchema: z.string(),
    streamSchema: z.string(),
  },
  async (type, { sendChunk }) => {
    const { stream, response } = ai.generateStream(
      `Tell me a longish ${type ?? "dad"} joke.`
    );
    for await (const chunk of stream) {
      sendChunk(chunk.text);
    }
    return (await response).text;
  }
);

export const tellJoke = onCallGenkit({ secrets: [apiKey] }, jokeTeller);
```

## Prompts

> We've make several improvements to prompts.

Reviewer: These are very cool features, but are the backward incompatible? If not, I'd suggest moving to a seperate "new features" section to make it clear which section of the doc is "you have to do this" vs "you might want to do this"

Reviewer 2: +1, I think it would help to break the guide into two sections:
- Namespace/syntax changes to keep existing code working
- New features since 0.9 (genkit express server, context, prompt enhancements)

Currently these are interwoven throughout.

Author: the prompts api is a breaking change....

You can define separate templates for prompt, system and messages:

> ```ts

Reviewer: The sentence says prompt, system and messages, but the example doesn't show messages

Author: There's no good practical example of using all 3 at the same time... I show messages down below

Reviewer: Then I'd drop messages from the sentence and include that in the messages bit

Author: good idea, done
